solutions/JOI 16-matryoshka.mdx

```diff
@@ -0,0 +1,100 @@
+---
+id: joi-16-matryoshka
+source: JOISC 2016
+title: Matryoshka
+author: Andi Qu
+---
+
+**Time Complexity**: $\mathcal O((N + Q) \log (\max H_i))$ or $\mathcal O((N + Q) \log N)$.
+
+First, let's consider a single query - what's the minimum number of dolls left over? A single nested doll consists of some dolls of increasing
+$R_i$ and $H_i$. If we sort the dolls by $H_i$ and only consider their $R_i$, then the answer is the minimum number of increasing subsequences
+needed to "cover" all the dolls.
+
+It's well-known that the minimum number of increasing subsequences is equal to the longest non-increasing subsequence (see [the LIS module](https://usaco.guide/gold/lis/#application-2---minimum-number-of-increasing-sequences)
+for more details.) This gives us a way to answer a single query in $\mathcal O(N \log N)$ time!
+
+But what if we need to answer multiple queries?
+
+Since we don't have to answer the queries online, let's sort them in decreasing order of $A_i$.
+If $B_i = \infty$ for each query, then we can simply do a line sweep on the dolls and queries,
+inserting dolls into the sequence and updating the longest non-increasing subsequence as we get to them.
+
+For variable $B_i$s, we only want the longest non-increasing subsequence for some prefixes
+of the sequence of dolls (not the whole sequence like before). We can thus use a segment tree
+to store, for each doll $i$, what that length is when considering only dolls with $H_j \leq H_i$.
+
+Since the coordinates can get very big, one must use either a sparse segment tree or coordinate
+compression to fit the segment tree in the memory. I used a sparse segment tree in my implementation below:
```

Comment:
no need for a segtree? just binary search on the LIS array.

Reply:
Should I still keep the segtree solution though?

Reply:
I don't see a reason to keep it (but you can if you want).